Model submission for Karen ground robot sensor configuration 1 and 2 from Coordinated Robotics

knoedler opened the pull request:
This is a four wheel robot with Ackermann type steering based on our "Karen" robot.

nkoenig commented:
Bounding box `[1.28749 0.765796 0.841283]`

acschang commented on submitted_models/coro_karen_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,1347 @@
+      <sensor name="camera_front" type="camera">
+        <pose>0.48 0.06 0.645 0 0 0</pose>
+        <always_on>1</always_on>
+        <update_rate>30</update_rate>
```
The specification sheet indicates shit sensor should be 25 Hz.

knoedler replied:
Fixed.

acschang commented on submitted_models/coro_karen_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,1347 @@
+      <sensor name="camera_rear" type="camera">
+        <pose>0.26 0 0.645 0 0 3.141592653589793</pose>
+        <always_on>1</always_on>
+        <update_rate>30</update_rate>
```
The specification sheet indicates shit sensor should be 25 Hz.

acschang commented on submitted_models/coro_karen_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,1347 @@
+      <sensor name="camera_left" type="camera">
+        <pose>0.37 0.11 0.645 0 0 1.5707963267948966</pose>
+        <always_on>1</always_on>
+        <update_rate>30</update_rate>
```
The specification sheet indicates shit sensor should be 25 Hz.

acschang commented on submitted_models/coro_karen_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,1347 @@
+      <sensor name="camera_right" type="camera">
+        <pose>0.37 -0.11 0.645 0 0 -1.5707963267948966</pose>
+        <always_on>1</always_on>
+        <update_rate>30</update_rate>
```
The specification sheet indicates shit sensor should be 25 Hz.

acschang reviewed (changes requested):
The initial assessment of the Coordinated Robotics Karen robot model is complete however there are some issues to resolve before this model can be merged and utilized in a competitive setting:

* Please add ROS to Ignition bridge nodes to the `vehicle_topics.launch` file for the barometer and magnetometer.
* Please add optical frame publishers for the Arducam camera streams in the `vehicle_topics.launch` file.
* The two RGB cameras in the `model.sdf` file do not match the Arducam specs. Was the intent to simulate a single 185deg-FOV Arducam with 2 cameras? If so, recommend adjusting framerate and combined resolution to match the Arducam specs (25Hz @ 1920x1080 resolution)
* Please provide validation data as it is necessary to complete the evaluation. If not provided, the model's maximum velocity will be limited to 1m/s and battery life set to 60min. When the model validation data can be provided later, please submit a PR updating the model.

Please address the issues noted above in this pull request. Once the above issues are fixed, your submission will move on to the next review stage!

knoedler commented on submitted_models/coro_karen_sensor_config_1/model.sdf:
```diff
@@ -0,0 +1,1347 @@
+          </image>
+          <clip>
+            <near>0.1</near>
+            <far>100</far>
```
Fixed

knoedler commented:
@acschang Thank you for reviewing.  Bridges and optical frame publishers were added.  I dropped the velocity limit to 1.0m/S until validation data is available and I believe the battery life is at 1 hour.  I adjusted the lidar range and camera rates / resolutions based on your comments.  I used 720x1080 for the camera resolutions.  The intent was to replace the two cameras with four cameras that provide 360 degree horizontal coverage and sensor matching vertical coverage.  I increased the vertical resolution to 1080 to match the sensor and adjusted the horizontal resolution to 720.  This gets the vertical FOV roughly correct and matches the real sensor as the full 1080 is usable, the full 1920 is not usable with that lens.  The description.launch now points at a urdf, but that urdf likely needs work (see #813)

acschang commented on submitted_models/coro_karen_sensor_config_1/launch/description.launch:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0"?>
+<launch>
+  <arg name="name" doc="Name of Vehicle"/>
+  <param name="$(arg name)/robot_description" command="$(find xacro)/xacro '$(find coro_karen_sensor_config_1)/urdf/karen_from_sdf.xacro' name:=$(arg name)"/>
```
I'm not able to visualize `RobotModel` type visualization in RViz.  this is the case on Allie as well while being related [the thread on Allie](https://github.com/osrf/subt/pull/813#discussion_r601572455).

knoedler replied:
Since the xacro / urdf is optional I went ahead and disabled it in the description.launch.  I will add it to our task list as an improvement that can be done later.

acschang commented on submitted_models/coro_karen_sensor_config_1/launch/spawner.rb:
```diff
@@ -0,0 +1,109 @@
+        <capacity>78.4</capacity>
+        <resistance>0.061523</resistance>
+        <smooth_current_tau>1.9499</smooth_current_tau>
+        <power_load>6.6</power_load>
```
These battery parameters are the same as the X1 which has a battery life of 90 minutes.  You can easily adjust this to 60 minutes by changing `power_load` to 9.9

knoedler replied:
Battery endurance fixed for config_1 and config_2, speed fixed for config_2

acschang commented:
> @acschang Thank you for reviewing. Bridges and optical frame publishers were added. I dropped the velocity limit to 1.0m/S until validation data is available and I believe the battery life is at 1 hour. I adjusted the lidar range and camera rates / resolutions based on your comments. I used 720x1080 for the camera resolutions. The intent was to replace the two cameras with four cameras that provide 360 degree horizontal coverage and sensor matching vertical coverage. I increased the vertical resolution to 1080 to match the sensor and adjusted the horizontal resolution to 720. This gets the vertical FOV roughly correct and matches the real sensor as the full 1080 is usable, the full 1920 is not usable with that lens. The description.launch now points at a urdf, but that urdf likely needs work (see #813)

Thank you for the changes.  It seems the only things outstanding for this model are the RViz `RobotModel` visualization and the battery endurance adjustment.

knoedler commented:
@acschang I believe the last two issues have been resolved.  The battery life has been fixed and the optional description.launch has been disabled.

nkoenig commented:
Continued in #865

nkoenig commented:
@knoedler , would you be able to add a model.urdf or model.xacro for this robot? This would let us have a `robot_description` present within ROS.

knoedler commented:
@nkoenig karen.xacro added in #873
